numpy/random/bit_generator.pyx

```diff
@@ -0,0 +1,613 @@
+
+cdef uint32_t mix(uint32_t x, uint32_t y):
+    cdef uint32_t result = (MIX_MULT_L * x - MIX_MULT_R * y)
+    result ^= result >> XSHIFT
+    return result
+
+
+class ISeedSequence(abc.ABC):
+    """
+    ISeedSequence() is the abstract base class for SeedSequences.
+
+    See Also
+    --------
+    SeedSequence, SeedlessSeedSequence
+    """
+
+    @abc.abstractmethod
+    def generate_state(self, n_words, dtype=np.uint32):
+        """
+        generate_state(n_words, dtype=np.uint32)
+
+        Return the requested number of words for PRNG seeding.
+
+        A BitGenerator should call this method in its constructor with
+        an appropriate `n_words` parameter to properly seed itself.
+
+        Parameters
+        ----------
+        n_words : int
+        dtype : np.uint32 or np.uint64, optional
+            The size of each word. This should only be either `uint32` or
+            `uint64`. Strings (`'uint32'`, `'uint64'`) are fine. Note that
+            requesting `uint64` will draw twice as many bits as `uint32` for
+            the same `n_words`. This is a convenience for `BitGenerator`s that
+            express their states as `uint64` arrays.
+
+        Returns
+        -------
+        state : uint32 or uint64 array, shape=(n_words,)
+        """
+
+
+class ISpawnableSeedSequence(ISeedSequence):
+
+    @abc.abstractmethod
+    def spawn(self, n_children):
+        """
+        spawn(n_children)
+
+        Spawn a number of child `SeedSequence` s by extending the
+        `spawn_key`.
+
+        Parameters
+        ----------
+        n_children : int
+
+        Returns
+        -------
+        seqs : list of `SeedSequence` s
+        """
+
+
+cdef class SeedlessSeedSequence():
+    """
+    SeedlessSeedSequence() is used for BitGenerators with no need for seed state.
+
+    See Also
+    --------
+    SeedSequence, ISeedSequence
+    """
+
+    def generate_state(self, n_words, dtype=np.uint32):
+        raise NotImplementedError('seedless SeedSequences cannot generate state')
+
+    def spawn(self, n_children):
+        return [self] * n_children
+
+
+# We cannot directly subclass a `cdef class` type from an `ABC` in Cython, so
+# we must register it after the fact.
+ISpawnableSeedSequence.register(SeedlessSeedSequence)
+
+
+cdef class SeedSequence():
+    """
+    SeedSequence(entropy=None, program_entropy=None, spawn_key=(), pool_size=4)
+
+    SeedSequence mixes sources of entropy in a reproducable way to set the
+    initial state for independent and very probably non-overlapping
+    BitGenerators.
+
+    Once the SeedSequence is instantiated, you can call the `generate_state`
+    method to get an appropriately sized seed. Calling `spawn(n) <spawn>` will
+    create ``n`` SeedSequences that can be used to seed independent
+    BitGenerators, i.e. for different threads.
+
+    Parameters
+    ----------
+    entropy : {None, int, sequence[int]}, optional
+        The entropy for creating a `SeedSequence`.
```

Comment:
I thought the point of SeedSequence is that `mix_entropy` mixes things up enough that even small integer seeds will now be "well behaved".

Reply:
I am still struggling with the whole "picking a good entropy". I would love to nudge users to good habits, and if just with documentation. E.g. saying: A good way to use as `entropy` is to use what `np.random.print_entropy()` prints out. The name `entropy` itself is a step up from `seed` though...

Reply:
There are multiple ways that small seeds can have bad effects. `SeedSequence` takes care of most of the ones that naive seeding might generate (e.g. mostly avoiding zero-land states from those PRNGs that are affected by that problem). But some are just intrinsic to small seeds. There are some possible initializations that you just won't be able to reach, and that can cause some biases over the ensemble of small-seed runs.

For many cases, that doesn't matter. If you just want to hold things in place while you debug something, biases aren't a concern.

For actual simulations whose results you care about, my recommendation is to let `SeedSequence(None)` do its thing and then log/print the `.entropy` that it sampled to initialize itself. That will be of a matching size to the pool. My advice if one wants to pick a seed "manually" "offline" is just to do `secrets.getrandbits(128)` (anyone mucking around with adjusting `pool_size` already knows what adjustment to make to that).

Reply:
Maybe a brief version of the second and third paragraphs of https://github.com/numpy/numpy/pull/13780#discussion_r296498703 could go in the documentation for the `numpy.random` module. It would succinctly answer the question most users will have: How do I pick a seed?

Reply:
expanded rationale and example use in 3b0cd7015